File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.cpp
#include <string>
#include <vector>
#include<stack>

using namespace std;

int solution(string s) {
    int answer = 0;
    for (int i = 0; i < s.length(); i++) {
	    stack<int> test;
        for (int j = i; j < i + s.length(); j++) {
            if (!test.empty()) {
                if ((s[j % s.length()] == ']' && test.top() == '[') ||
                    (s[j % s.length()] == '}' && test.top() == '{') ||
                    (s[j % s.length()] == ')' && test.top() == '(')) test.pop();
                else test.push(s[j % s.length()]);
            }
            else test.push(s[j % s.length()]);
        }
	    if(test.empty()) answer++;
    }
    return answer;
}

File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.java
import java.util.*;

class Solution {
    static boolean check(String s){
        Stack<Character> st = new Stack<>();
        
        for(int i=0; i<s.length(); i++){
            char c = s.charAt(i);
            if(c=='(' || c=='[' || c=='{') st.push(c);
            else if(!st.empty()){
                if((c==')' && st.peek()=='(') ||
                   (c==']' && st.peek()=='[') ||
                   (c=='}' && st.peek()=='{')) st.pop();
            }
            else return false;
        }
        return st.empty();
    }
    
    public int solution(String s) {
        int answer=0;
        for(int i=0; i<s.length(); i++){
            String rotate = s.substring(i) + s.substring(0,i);
            if(check(rotate)) answer++;
        }
        return answer;
    }
}